runtime/dart_isolate.cc

```diff
@@ -639,6 +639,7 @@ Dart_Isolate DartIsolate::DartIsolateGroupCreateCallback(
     Dart_IsolateFlags* flags,
     std::shared_ptr<DartIsolate>* parent_embedder_isolate,
     char** error) {
+  TRACE_EVENT0("flutter", "DartIsolate::DartIsolateGroupCreateCallback");
```

Comment:
(I cannot comment on the line blow due to github review, but this comment belongs to line 664)

The code below should be unreachable or should error if there is no support for `Isolate.spawnUri`, since every `Isolate.spawn` will cause us to use the `initialize_isolate` callback. The only other isolate groups created are for service/kernel isolate.

That being said, I don't know how the "multiple flutter engine shell" feature works, though I assume they will not use this code either.

Therefore I think the call to `CreateDartVMAndEmbedderObjectPair(..., is_root_isolate=false)` should be replaced with an error (or unreachable).

Reply:
Not sure I understand. This `DartIsolateGroupCreateCallback` is called on isolate group creation. That will be called two times(service isolate group, main isolate group) and for main isolate the call `CreateDartVMAndEmbedderObjectPair(..., is_root_isolate=false)` will be triggered, right?

Reply:
The `DartIsolateGroupCreateCallback` is called **from the VM**, if the **VM wants to create new isolate groups**. This happens in these cases:

   * VM wants to create `service-isolate`
   * `Isolate.spawn` (**only** if isolate groups are disabled,  if they are enabled then only `initialize_isolate` callback is used, not `create_group`)
   * `Isolate.spawnUri` (not supported by flutter I suppose)
   * VM wants to create `kernel-isolate` (not used by flutter I suppose)

=> If isolate groups are enabled, the only thing this callback will ever get to do is starting the service isolate.

The actual main UI isolate is created by the embedder itself and doesn't go through any callbacks. I believe this is done in [runtime/runtime_controller.cc](https://github.com/flutter/engine/blob/master/runtime/runtime_controller.cc#L75)


That being said, I think you can ignore my original comment: As discussed offline we want to have the flexibility to flip `--enable-isolate-groups` on/off for some time, so we keep this code.